### ISSUE TITLE
Add Weight Decay to optimizers

### ISSUE THREAD

**Issue by coreylowman** [coreylowman/dfdx#99]
- [x] Sgd `weight_decay: Option<f32>`
- [ ] RMSprop `weight_decay: Option<f32>`
- [ ] Adam/AdamW: `weight_decay: Option<AdamDecay>`, where `enum AdamDecay { Vanilla(f32), AdamW(f32) }`.

(maybe use different terms for `AdamDecay`).

**Comment by cBournhonesque** [comment]
Hi, I wanted to work on this issue but I think I need some guidance.

1. If I understand correctly, the gradients returned by [this function](https://github.com/coreylowman/dfdx/blob/main/src/optim/sgd.rs#L120) are separate from the gradients present in the tensors' tape, and are safe to be modified? i.e. modifying the gradients inside this function won't modify the gradients associated with each parameter weight, they only represent the 'update' to be passed to each tensor during the optimization step.

2. Provided 1. is true and I can modify `*g` inside that function directly, the only thing I need to do is to add some code like
```
if let Some(wd) = self.cfg.weight_decay {
            P::Device::foreach_m(g_t.as_mut(), &mut |g| *g += wd * param_weights )
}
```
above [this line](https://github.com/coreylowman/dfdx/blob/main/src/optim/sgd.rs#L142) ?

3. The other issue is that I would need to have access to the param weights inside the sgd `gradient` function. It seems like there needs to be a function `weights` in the `Module` trait that gives me access to all the param weights of a module? (an object similar to `Gradients`, something like a HashMap<id, Array>)

**Comment by coreylowman** [comment]
1. Yeah they are separate from gradients inside the tape. You can kind of think about optimizers as a mapping function that take the raw gradient from the tape, apply some functions to it, and then return it so it can be used to update the tensor.
2. Yep! But double check which line exactly should get the weight decay update. The pytorch docs have good descriptions for the optimizers. For example in [SGD](https://pytorch.org/docs/stable/generated/torch.optim.SGD.html) they show the weight decay operation happening before the momentum updates
3. The `p: &P` parameter passed into gradient function is the parameter! So you can call `p.data()` to get a reference to the array.

Also don't feel the need to add the functionality to all optimizers before opening a PR. Will be easier to review if they are added 1 at a time. Probably should've split this issue into 3 separate ones in retrospect haha

**Comment by cBournhonesque** [comment]
Another thing, I'm having a bit of trouble with the tests, because of numerical instability of how floats are represented?
If I try to compute the expected results in python, I get slightly different results at the high decimals. What do you advise?

**Comment by coreylowman** [comment]
Ahh yeah we'll need to add HasArrayData to the GradientProvider trait and the existing optimizers.

The L2 vs decoupled is just for Adam right? That's why I was thinking we could have an enum with the two options instead of just an Option<f32>:
```rust
enum AdamWeightDecay {
    L2(f32),
    Decoupled(f32),
}
```

Then we could have in Adam's code:
```rust
if let Some(AdamWeightDecay::L2(w)) = self.weight_decay { ... }
...
if let Some(AdamWeightDecay::Decoupled(w)) = self.weight_decay { ... }
```

I like the enum because then there's no confusion on user's PoV which version is being used (they don't need to check documentation). However it does mean user would have to learn the difference between the two methods (maybe a good thing?).

**Comment by coreylowman** [comment]
> Another thing, I'm having a bit of trouble with the tests, because of numerical instability of how floats are represented? If I try to compute the expected results in python, I get slightly different results at the high decimals. What do you advise?

Will reply in PR

**Comment by cBournhonesque** [comment]
> Ahh yeah we'll need to add HasArrayData to the GradientProvider trait and the existing optimizers.
> 
> The L2 vs decoupled is just for Adam right? That's why I was thinking we could have an enum with the two options instead of just an Option:
> 
> ```rust
> enum AdamWeightDecay {
>     L2(f32),
>     Decoupled(f32),
> }
> ```
> 
> Then we could have in Adam's code:
> 
> ```rust
> if let Some(AdamWeightDecay::L2(w)) = self.weight_decay { ... }
> ...
> if let Some(AdamWeightDecay::Decoupled(w)) = self.weight_decay { ... }
> ```
> 
> I like the enum because then there's no confusion on user's PoV which version is being used (they don't need to check documentation). However it does mean user would have to learn the difference between the two methods (maybe a good thing?).

Seems good to me!
It's not only for Adam, it's for any optimizer that has momentum. There's also [SGDW](https://www.tensorflow.org/addons/api_docs/python/tfa/optimizers/SGDW) which makes the same distinction (decay applied after momentum)

**Comment by coreylowman** [comment]
Okay yeah then I think we could just add a `enum WeightDecay` that all of them use in that case (no need to call it AdamWeightDecay)!
